openedx/core/djangoapps/programs/utils.py

```diff
@@ -151,7 +193,12 @@ def progress(self, programs=None, count_only=True):
                 if self._is_course_complete(course):
                     completed.append(course)
                 elif self._is_course_in_progress(course):
-                    in_progress.append(course)
+                    course_in_progress = self.does_course_still_qualify_as_in_progress(now, enrolled_run_modes, course)
+                    if course_in_progress:
+                        in_progress.append(course)
+                    else:
+                        course['expired'] = True
```

Comment:
To clarify, is your suggestion to add this `expired` key to a copy of the course that's returned in the progress dict? If so, I think that's fine. We just shouldn't modify any programs as a side effect of calling this method.